### PR TITLE
Add full kotlin configurations to kotlin rule

### PR DESCRIPTION
**Pull request by ZacSweers**
These match the configurable JVM arguments in other build systems and better match naming conventions.

Note that much of this logic is borrowed from OkBuck, where we already extract these arguments from gradle configurations

CC @thalescm @kageiit @raviagarwal7

### CODE REVIEW

**Comment by hzsweers** [comment]
Note that I'm not very familiar with the project and am not sure if/where to add tests, so guidance welcome!

File: src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java [hunk 1]
```diff
@@ -193,8 +193,38 @@ public void findDepsForTargetFromConstructorArgs(
...
   public interface CoreArg extends JavaLibraryDescription.CoreArg {
+    /**
+     * @deprecated use {@link #getFreeCompilerArgs()}.
+     */
+    @Deprecated
     ImmutableList<String> getExtraKotlincArguments();
 
+    ImmutableList<String> getFreeCompilerArgs();
+
+    Optional<Boolean> getAllWarningsAsErrors();
```
Comment (raviagarwal7) [inline src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java]:
```suggestion
    @Value.Default
    default boolean getAllWarningsAsErrors() {
      return false;
    }    
```

you can provide defaults for all booleans

Comment (hzsweers) [reply src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java]: Does `@Value.Default` ensure that an implementation still gets generated under the hood too?

Comment (raviagarwal7) [reply src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java]: yes, it does create an implementation which fallbacks to the default implementation if not set.

File: src/com/facebook/buck/jvm/kotlin/KotlinConfiguredCompilerFactory.java [hunk 2]
```diff
@@ -73,4 +85,69 @@ public CompileToJarStepFactory configure(
   private Javac getJavac(BuildRuleResolver resolver, @Nullable JvmLibraryArg arg) {
     return javacFactory.create(new SourcePathRuleFinder(resolver), arg);
   }
+
+  private ImmutableList<String> condenseCompilerArguments(CoreArg kotlinArgs) {
```
Comment (raviagarwal7) [inline src/com/facebook/buck/jvm/kotlin/KotlinConfiguredCompilerFactory.java]: seems like there is no existing test for `KotlinConfiguredCompilerFactory.java`. you can probably add one for this method.

File: src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java (hunk 1 above)
Comment (hzsweers) [reply src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java]: 02d8940

File: src/com/facebook/buck/jvm/kotlin/KotlinConfiguredCompilerFactory.java (hunk 2 above)
Comment (hzsweers) [reply src/com/facebook/buck/jvm/kotlin/KotlinConfiguredCompilerFactory.java]: any good prior examples you could point me to?

File: src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java [hunk 3]
```diff
@@ -193,8 +193,65 @@ public void findDepsForTargetFromConstructorArgs(
...
+
+    @Value.Default
+    default boolean getNoReflect() {
+      return true;
+    }
+
+    @Value.Default
+    default boolean getJavaParameters() {
+      return false;
+    }
```
Comment (raviagarwal7) [inline src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java]: also add a section in https://github.com/facebook/buck/blob/master/docs/rule/kotlin_library.soy describing these parameters.

File: src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java [hunk 4]
```diff
@@ -193,8 +193,65 @@ public void findDepsForTargetFromConstructorArgs(
...
+
+    @Value.Default
+    default boolean getSuppressWarnings() {
+      return false;
+    }
+
+    @Value.Default
+    default boolean getVerbose() {
+      return false;
```
Comment (raviagarwal7) [inline src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java]: seems like verbose should be true by default as per the usage in below `KotlincStep`

File: src/com/facebook/buck/jvm/kotlin/KotlinConfiguredCompilerFactory.java (hunk 2 above)
Comment (raviagarwal7) [reply src/com/facebook/buck/jvm/kotlin/KotlinConfiguredCompilerFactory.java]: checkout https://github.com/facebook/buck/blob/master/test/com/facebook/buck/jvm/kotlin/KotlinBuckConfigTest.java for unit tests example

for integration tests https://github.com/facebook/buck/blob/master/test/com/facebook/buck/jvm/kotlin/KotlinLibraryIntegrationTest.java

there are bunch of test data https://github.com/facebook/buck/tree/master/test/com/facebook/buck/jvm/kotlin/testdata, you can probably add an integration test too.

File: src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java [hunk 5]
```diff
@@ -223,11 +220,8 @@ public void createCompileStep(
               kotlinc,
               ImmutableList.<String>builder()
                   .addAll(extraKotlincArguments)
-                  .add(NO_STDLIB)
-                  .add(NO_REFLECT)
                   .add(COMPILER_BUILTINS)
                   .add(LOAD_BUILTINS_FROM)
```
[the remark this answers is not in the document]
Comment (thalescm) [reply src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java]: Also, the `condenseCompilerArguments` would remove duplicates for this if user also added them :)

File: src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java (hunk 4 above)
Comment (hzsweers) [reply src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java]: I think that was incorrect before, as the default is `false`

File: src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (hzsweers) [reply src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java]: True, I can move these as well. I admittedly am not familiar with these args

**Comment by hzsweers** [comment]
The CI failures look unrelated

File: src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (thalescm) [reply src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java]: I think it was me (or @cwoodwar6 ) who added them in order to get kapt to work. This was actually a copy past from what gradle outputs when doing the same.
Also, could you also remove them from [here](https://github.com/facebook/buck/blob/master/src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java#L359-L360) and [here](https://github.com/facebook/buck/blob/master/src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java#L387-L388) when adding them to `KotlinConfiguredCompilerFactory`?

[the remark this answers is not in the document]
Comment (ZacSweers) [reply src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java]: Consolidated in b48992c

[the remark this answers is not in the document]
Comment (ZacSweers) [reply src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java]: It actually looks like these two are false by default and not exactly documented. I'm not sure if they're maybe just required for kapt, so going to leave them as-is for now rather than move to the standard options. If they're not required for kapt either, we can just look at getting rid of them

**Comment by ZacSweers** [comment]
I've taken a full documentation pass an opted to just remove the old style for args. Before I move on to tests, I'd like to get another review of the current implementation (particularly from the buck folks).

The CI failures look unrelated. I've noticed them on a few other PRs as well

**Comment by ZacSweers** [comment]
image

**Comment by kageiit** [comment]
@ZacSweers can you please rebase this?

**Comment by kageiit** [comment]
Lets make sure to run gjf as well upon rebase so we can shorten review cycles

**Comment by ZacSweers** [comment]
I've rebased and run GJF over the files

**Comment by kageiit** [comment]
Seems like build started failing after rebase (compilation, not tests)

**Comment by ZacSweers** [comment]
Hard to keep track, as it was failing due to issues on master before I rebased too. I don't really know what's going on with it. Some sort of interface merging issue in Immutables? `setLanguageVersion` isn't part of the android library description, so I don't know why it's pointing there

```
java.lang.IllegalStateException: Couldn't find declared getter for class com.facebook.buck.android.AndroidLibraryDescriptionArg$Builder#setLanguageVersion. Tried parent classes [interface com.facebook.buck.android.AndroidLibraryDescription$AbstractAndroidLibraryDescriptionArg, class java.lang.Object, interface com.facebook.buck.android.AndroidLibraryDescription$CoreArg, interface com.facebook.buck.jvm.java.JavaLibraryDescription$CoreArg, interface com.facebook.buck.android.AndroidKotlinCoreArg, interface com.facebook.buck.core.description.arg.HasDepsQuery, interface com.facebook.buck.core.description.arg.HasProvidedDepsQuery, interface com.facebook.buck.jvm.java.JvmLibraryArg, interface com.facebook.buck.core.description.arg.HasDeclaredDeps, interface com.facebook.buck.core.description.arg.HasProvidedDeps, interface com.facebook.buck.core.description.arg.HasSrcs, interface com.facebook.buck.core.description.arg.HasTests, interface com.facebook.buck.jvm.kotlin.KotlinLibraryDescription$CoreArg, interface com.facebook.buck.core.description.arg.CommonDescriptionArg, interface com.facebook.buck.jvm.java.MaybeRequiredForSourceOnlyAbiArg, interface com.facebook.buck.core.description.arg.HasTargetCompatibleWith] methods: [getLanguageVersion, isLanguageVersion]
	at com.facebook.buck.rules.coercer.ParamInfo.findClosestGetterOnAbstractClassOrInterface(ParamInfo.java:317)
	at com.facebook.buck.util.function.ThrowingSupplier.lambda$1(ThrowingSupplier.java:41)
	at com.facebook.buck.util.AbstractMemoizer.get(AbstractMemoizer.java:42)
```

**Comment by ZacSweers** [comment]
Abandoning this on account of lack of feedback/interest from the BUCK team.

File: docs/rule/kotlin_library.soy [hunk 6]
```diff
@@ -69,10 +69,109 @@ the <code>resources</code> argument.
 {/call}
 
 {call buck.arg}
-  {param name: 'extra_arguments' /}
+  {param name: 'free_compiler_args' /}
```
Comment (styurin) [inline docs/rule/kotlin_library.soy]: Why not `compiler_args`? Even `extra_compiler_args`. It's not what "free" means here and we use `extra` in some other rules.

File: src/com/facebook/buck/jvm/kotlin/KotlinConfiguredCompilerFactory.java [hunk 7]
```diff
@@ -74,4 +86,70 @@ public CompileToJarStepFactory configure(
...
+              if (entry.getValue().isPresent()) {
+                return ImmutableList.of(entry.getKey(), entry.getValue().get()).stream();
+              } else {
+                return ImmutableList.of(entry.getKey()).stream();
+              }
+            })
+        .collect(toImmutableList());
+  }
+
+  private static <T> Predicate<T> distinctByKey(Function<? super T, ?> keyExtractor) {
```
Comment (styurin) [inline src/com/facebook/buck/jvm/kotlin/KotlinConfiguredCompilerFactory.java]: Can you add a test to make sure duplicate arguments are handled properly?

File: docs/rule/kotlin_library.soy (hunk 6 above)
Comment (ZacSweers) [reply docs/rule/kotlin_library.soy]: I was naming these to match the naming semantics that the kotlin tooling itself uses for consistency

**Comment by ZacSweers** [comment]
@styurin I'm closing this as review has stalled again and I'd have to chase rebase conflicts again because of it.  @kageiit and @thalescm have push access to my fork if you all are genuinely interested in pursuing this, but I'm afraid I can't afford to sink any more time into this. I have another branch on my fork focused on adding kotlin facet generation to intellij project imls, something that would be a really significant QoL improvement to kotlin consumers since much of the IDE integration depends on being able to read this metadata from the build configuration via these facets.

On a meta note - I understand that kotlin isn't a primary focus of the buck team, but I don't think that's the reason this PR has gone the way it has. There are nearly 40 open PRs and over 10 times as many open issues. Most of them almost certainly are not kotlin-related. Some are merged internally but never cleared up here, many are just not acknowledged, and seems to reflect the "disclosed source" nature of this project rather than "open source". Disclosed source is perfectly ok! No one's entitled to have their code merged and there are a lot of great projects that follow this pattern (AOSP itself is a primary example). However, I think there's clearly some room for improvement as far as setting expectations and engaging community contributions in this project, and would encourage the buck team to proactively engage this rather than maintain the status quo.
